### ISSUE TITLE
Spike:  Kafka trigger for a Lambda POC

### ISSUE THREAD

**Issue by nikolai-efimov** [department-of-veterans-affairs/notification-api#1369]
## User Story - Business Goal
As we architect this new process for Comp and Pen to consume payment events and then kick off subsequent processing actions, we want to ensure we are doing so in the most performant, efficient, cost effective way using the tools we have available to us. We had previously been discussing using EC2, but have concerns about the LOE to set this up and the fact that it doesn't need to be running all the time. We know these batches will only come in once a month and we will likely only need to process these over one or a few days.

- [ ] Ticket is understood, and QA has been contacted (if the ticket has a QA label).

### User Story(ies)

**As** VA Notify
**I want** to determine if a notification should be sent when I receive a payment event from BIP
**So that** I'm only sending SMS to Veterans who have opted in

### Additional Info and Resources
- https://docs.aws.amazon.com/lambda/latest/dg/invocation-eventsourcemapping.html

## Acceptance Criteria & Checklist

- [x] Verified that feature is available in gov space (kafka trigger in lambda)
- [ ] Trigger was successfully attached to a lambda
- [ ] Messages are received and decoded (use code from [Spike 1363](https://app.zenhub.com/workspaces/va-notify-620d21369d810a00146ed9c8/issues/gh/department-of-veterans-affairs/notification-api/1363)
- [ ] Set batch size to 100 and verify that if needed, we can invoke even if there are only 99 messages
- [ ] Verified that it can be setup using Terraform. 
- [ ] Technical Documentation has been created and linked in the glossary
- [ ] Learnings have been shared with the team, and decisions have been documented
- [ ] Tickets for next steps have been identified and created

Completed research addresses the problems and/or answers the questions below:  

1. Is the feature available in gov space? Because of compliance requirements, it previously was only available for non-gov accounts.
2. Can we trigger a lambda using BIP's topic?
3. Can we decode messages properly using Avro Schema Registry? (refer to [Spike 1363](https://app.zenhub.com/workspaces/va-notify-620d21369d810a00146ed9c8/issues/gh/department-of-veterans-affairs/notification-api/1363))
4. Is there a 'batch window timeout' for Kafka? i.e. how to handle a situation where our batch size is set to 100, but there are only 1-99 messages in there.

## QA Considerations
- [ ] Sync with assignee to understand how this will work

**Comment by mjones-oddball** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/VA-Notify-620d21369d810a00146ed9c8/issues/department-of-veterans-affairs/notification-api/1369?planning-poker) with Zenhub @cris-oddball @EvanParish @justaskdavidb2 @k-macmillan @kalbfled @ldraney @nikolai-efimov

**Comment by mjones-oddball** [comment]
Kyle verified he can see kafka as a trigger in the lambda console

**Comment by cris-oddball** [comment]
Moved out of sprint 40 into backlog as a stretch goal if the rest of sprint 40 is completed.

**Comment by nikolai-efimov** [comment]
![Screen Shot 2023-08-15 at 11.35.49 AM.png](https://images.zenhubusercontent.com/647df92fea2b08c16fcd5059/fffd85cf-1808-4749-b2b5-9811e0c3dfee)

**Comment by nikolai-efimov** [comment]
![Screen Shot 2023-08-15 at 11.42.10 AM.png](https://images.zenhubusercontent.com/647df92fea2b08c16fcd5059/8b12f8ad-5c87-440f-93d3-bf064b1ad969)

**Comment by tabinda-syed** [comment]
Increased to a 5 based on Evan's assessment after his research on 8/17.

**Comment by EvanParish** [comment]
We have tried a couple different ways of getting a lambda up to our aws account for testing without deploying terraform, but all of them failed. I created some terraform that should get us most of the way there, and hopefully we can manually adjust anything necessary to get it fully working.

**Comment by EvanParish** [comment]
Here's a [link to the PR](https://github.com/department-of-veterans-affairs/vanotify-infra/pull/630) for the code I've worked on so far.

**Comment by EvanParish** [comment]
I finally got the placeholder lambda and kafka trigger in place! Now I *just* need to get the certificates and code in place to test it. I also need to figure out how to consumer the incoming messages exactly. 
Another big question.
- Can we read from the beginning of the stream with the trigger set up like this?

**Comment by EvanParish** [comment]
Progress made on this ticket:
- it seems that our terraform aws provider version may be too old to properly implement the self-managed kafka trigger
  - the trigger can be configured manually in the console once the lambda is in place
- after a lot of trial and error with certs, I think we found the proper format for the kafka trigger to accept them from the secrets manager
- the lambda has yet to be successfully triggered by the self-managed kafka event source
  - we are now doing testing with our own producer in confuent cloud in a [new ticket team-1125](https://app.zenhub.com/workspaces/va-notify-620d21369d810a00146ed9c8/issues/gh/department-of-veterans-affairs/vanotify-team/1125)

@nikolai-efimov is there anything else you'd like to add? 
Don't forget that the terraform that was created so far can be found in this [draft PR](https://github.com/department-of-veterans-affairs/vanotify-infra/pull/630) in case it need to be redeploy. 
branch name: 1369-kafka-trigger-for-lambda-poc
